lib/IDE/CodeCompletion.cpp

```diff
@@ -935,24 +934,35 @@ calculateTypeRelationForDecl(const Decl *D, Type ExpectedType,
     return CodeCompletionResult::ExpectedTypeRelation::Unrelated;
 
   if (auto FD = dyn_cast<AbstractFunctionDecl>(VD)) {
-    auto funcType = FD->getType()->getAs<AnyFunctionType>();
-    if (DC->isTypeContext() && funcType && funcType->is<AnyFunctionType>() &&
-        !IsImplicitlyCurriedInstanceMethod)
-      funcType = funcType->getResult()->getAs<AnyFunctionType>();
+    auto funcType = FD->getInterfaceType()->getAs<AnyFunctionType>();
+    // Strip the implicit self parameter off of a function type.
+    if (DC->isTypeContext() && funcType && !IsImplicitlyCurriedInstanceMethod) {
+      auto newFuncType = funcType->getResult()->getAs<AnyFunctionType>();
+      if (auto *genFuncType = dyn_cast<GenericFunctionType>(funcType)) {
+        if (newFuncType) {
+          // Re-wrap in GenericFunctionType to maintain the GenericSignature.
+          newFuncType = GenericFunctionType::get(
```

Comment:
It's unfortunate that we're stripping off the 'self' type here entirely, when it's likely that we actually have a 'self' type we can use and could be substituted into the function type.